### PR TITLE
Reuse hash

### LINKED ISSUE #11680: Improve performance of high cardinality grouping by reusing hash values
### Is your feature request related to a problem or challenge?

As described on https://github.com/apache/datafusion/issues/11679, we can do better for high cardinality aggregates

One thing that consumes significant time in such queries is hashing, and I think we can reduce that significantly.

Specifically, for the multi-phase repartition plan, the number of hashed rows is something like 

```
(input cardinality)  + 2 * (intermediate group cardinality) * (number of partitions)
```

For low cardinality aggregates (e.g when the intermediate group cardinality is 1000) the second term is small (a few thousand extra hashes isn't a big deal)

However, for high cardinality aggregates (eg. when the intermediate cardinality is like 1,000,000 and there are 16 partitions) the second term is substantial

In pictures, this looks like

```
               ▲                          ▲
               │                          │
               │                          │
               │                          │
               │                          │
               │                          │
   ┌───────────────────────┐  ┌───────────────────────┐       4. The  AggregateMode::Final
   │GroupBy                │  │GroupBy                │       GroupBy computes hash(group keys)
   │(AggregateMode::Final) │  │(AggregateMode::Final) │       *AGAIN* to find the correct hash
   │                       │  │                       │       bucket
   └───────────────────────┘  └───────────────────────┘
               ▲                          ▲
               │                          │
               └─────────────┬────────────┘
                             │
                             │
                             │
                ┌─────────────────────────┐                   3. The output of the first phase
                │       Repartition       │                   is repartitioned by computing
                │         HASH(x)         │                   hash(group keys) -- this is the
                └─────────────────────────┘                   same hash as computed in step 2.
                             ▲
                             │
             ┌───────────────┴─────────────┐
             │                             │
             │                             │
┌─────────────────────────┐  ┌──────────────────────────┐     2. Each AggregateMode::Partial
│        GroubyBy         │  │         GroubyBy         │     GroupBy hashes the group keys to
│(AggregateMode::Partial) │  │ (AggregateMode::Partial) │     find the correct hash bucket.
└─────────────────────────┘  └──────────────────────────┘
             ▲                             ▲
             │                            ┌┘
             │                            │
        .─────────.                  .─────────.
     ,─'           '─.            ,─'           '─.
    ;      Input      :          ;      Input      :          1. Input is read
    :   Partition 0   ;          :   Partition 1   ;
     ╲               ╱            ╲               ╱
      '─.         ,─'              '─.         ,─'
         `───────'                    `───────'
```

This effect can be seen in profiling for click bench XXX:

(TODO screen shot showing large amounts of work done in hashing)

### Describe the solution you'd like

The basic idea is rather than rather than recompute the hash values in `RepartitionExec` and `AggregateMode::Final` we would reuse the values from `AggregateMode::Partial` (which has already computed a hash value for each input group)

Something like this
```text
                         ▲                          ▲                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
             ┌───────────────────────┐  ┌───────────────────────┐       4. The  AggregateMode::Final        
             │GroupBy                │  │GroupBy                │       GroupBy also gets the hash values   
             │(AggregateMode::Final) │  │(AggregateMode::Final) │       and does not recompute them         
             │                       │  │                       │                                           
             └───────────────────────┘  └───────────────────────┘                                           
               ▲         ▲                          ▲                                                       
               │         │                          │                                                       
                         └─────────────┬────────────┘                                                       
Pass hash      │                       │                                                                    
values up the                          │                                                                    
plan rather    │                       │                                                                    
than                      ┌─────────────────────────┐                   3. In addition to the partial       
recomputing    │          │       Repartition       │                   aggregates and group values, *ALSO* 
them                      │    PRECOMPUTED_HASH     │                   pass the hash values to the         
               │          └─────────────────────────┘                   RepartitionExec which also passed   
                                       ▲                                them on to the AggregateMode::Final 
               │                       │                                                                    
                       ┌───────────────┴─────────────┐                                                      
               │       │                             │                                                      
                       │                             │                                                      
          ┌─────────────────────────┐  ┌──────────────────────────┐     2. Each AggregateMode::Partial      
          │        GroubyBy         │  │         GroubyBy         │     GroupBy hashes the group keys to    
          │(AggregateMode::Partial) │  │ (AggregateMode::Partial) │     find the correct hash bucket.       
          └─────────────────────────┘  └──────────────────────────┘                                         
                       ▲                             ▲                                                      
                       │                            ┌┘                                                      
                       │                            │                                                       
                  .─────────.                  .─────────.                                                  
               ,─'           '─.            ,─'           '─.                                               
              ;      Input      :          ;      Input      :          1. Input is read                    
              :   Partition 0   ;          :   Partition 1   ;                                              
               ╲               ╱            ╲               ╱                                               
                '─.         ,─'              '─.         ,─'                                                
                   `───────'                    `───────'                                                   
```

### Describe alternatives you've considered

We maybe could pass the data as an explicit new column somehow, or maybe as a field in a struct array 🤔 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #11680

### CODE REVIEW

File: benchmarks/queries/clickbench/queries.sql [hunk 1]
```diff
@@ -1,43 +1,3 @@
...
-SELECT "RegionID", SUM("AdvEngineID"), COUNT(*) AS c, AVG("ResolutionWidth"), COUNT(DISTINCT "UserID") FROM hits GROUP BY "RegionID" ORDER BY c DESC LIMIT 10;
-SELECT "MobilePhoneModel", COUNT(DISTINCT "UserID") AS u FROM hits WHERE "MobilePhoneModel" <> '' GROUP BY "MobilePhoneModel" ORDER BY u DESC LIMIT 10;
-SELECT "MobilePhone", "MobilePhoneModel", COUNT(DISTINCT "UserID") AS u FROM hits WHERE "MobilePhoneModel" <> '' GROUP BY "MobilePhone", "MobilePhoneModel" ORDER BY u DESC LIMIT 10;
-SELECT "SearchPhrase", COUNT(*) AS c FROM hits WHERE "SearchPhrase" <> '' GROUP BY "SearchPhrase" ORDER BY c DESC LIMIT 10;
-SELECT "SearchPhrase", COUNT(DISTINCT "UserID") AS u FROM hits WHERE "SearchPhrase" <> '' GROUP BY "SearchPhrase" ORDER BY u DESC LIMIT 10;
-SELECT "SearchEngineID", "SearchPhrase", COUNT(*) AS c FROM hits WHERE "SearchPhrase" <> '' GROUP BY "SearchEngineID", "SearchPhrase" ORDER BY c DESC LIMIT 10;
-SELECT "UserID", COUNT(*) FROM hits GROUP BY "UserID" ORDER BY COUNT(*) DESC LIMIT 10;
 SELECT "UserID", "SearchPhrase", COUNT(*) FROM hits GROUP BY "UserID", "SearchPhrase" ORDER BY COUNT(*) DESC LIMIT 10;
```
Comment (jayzhan211) [inline benchmarks/queries/clickbench/queries.sql:1]: no change compared to main

File: benchmarks/queries/clickbench/queries.sql [hunk 2]
```diff
@@ -1,43 +1,3 @@
...
-SELECT "WindowClientWidth", "WindowClientHeight", COUNT(*) AS PageViews FROM hits WHERE "CounterID" = 62 AND "EventDate"::INT::DATE >= '2013-07-01' AND "EventDate"::INT::DATE <= '2013-07-31' AND "IsRefresh" = 0 AND "DontCountHits" = 0 AND "URLHash" = 2868770270353813622 GROUP BY "WindowClientWidth", "WindowClientHeight" ORDER BY PageViews DESC LIMIT 10 OFFSET 10000;
-SELECT DATE_TRUNC('minute', to_timestamp_seconds("EventTime")) AS M, COUNT(*) AS PageViews FROM hits WHERE "CounterID" = 62 AND "EventDate"::INT::DATE >= '2013-07-14' AND "EventDate"::INT::DATE <= '2013-07-15' AND "IsRefresh" = 0 AND "DontCountHits" = 0 GROUP BY DATE_TRUNC('minute', to_timestamp_seconds("EventTime")) ORDER BY DATE_TRUNC('minute', M) LIMIT 10 OFFSET 1000;
+SELECT "UserID", concat("SearchPhrase", repeat('hello', 100)) as s, COUNT(*) FROM hits GROUP BY "UserID", s ORDER BY COUNT(*) DESC;
```
Comment (jayzhan211) [inline benchmarks/queries/clickbench/queries.sql:2]: 2x slower

File: datafusion/physical-plan/src/aggregates/group_values/row.rs [hunk 3]
```diff
@@ -111,45 +123,90 @@ impl GroupValues for GroupValuesRows {
...
-                    self.map.insert_accounted(
-                        (hash, group_idx),
-                        |(hash, _group_index)| *hash,
-                        &mut self.map_size,
-                    );
-                    group_idx
-                }
-            };
-            groups.push(group_idx);
+        // let hash_values: Option<&ArrayRef> = None;
```
Comment (jayzhan211) [inline datafusion/physical-plan/src/aggregates/group_values/row.rs:134]: The most surprised things is I got the similar result as main if I uncomment this line...

The only difference is one iterate U64Array, another calculate the hashes and get `Vec<u64>` directly

**Comment by jayzhan211** [comment]
Screenshot 2024-07-30 at 11 12 28 AM

The time of `set_input_done_and_produce_output` grows to over 1.29s, it is 555ms in `main`. Not clear what is the reason

**Comment by jayzhan211** [comment]
Screenshot 2024-07-30 at 5 54 04 PM

File: datafusion/physical-plan/src/aggregates/group_values/row.rs [hunk 4]
```diff
@@ -91,15 +97,21 @@ impl GroupValuesRows {
             map,
             map_size: 0,
             group_values: None,
+            group_hashes: Default::default(),
             hashes_buffer: Default::default(),
             rows_buffer,
-            random_state: Default::default(),
+            random_state: RandomState::with_seeds(0, 0, 0, 0),
```
Comment (jayzhan211) [inline datafusion/physical-plan/src/aggregates/group_values/row.rs:103]: With fixed seed to ensure the hash value is the same for the same group values

File: datafusion/physical-plan/src/repartition/mod.rs [hunk 5]
```diff
@@ -256,30 +255,25 @@ impl BatchPartitioner {
...
-                    create_hashes(&arrays, random_state, hash_buffer)?;
-
                     let mut indices: Vec<_> = (0..*partitions)
                         .map(|_| UInt64Builder::with_capacity(batch.num_rows()))
                         .collect();
 
-                    for (index, hash) in hash_buffer.iter().enumerate() {
-                        indices[(*hash % *partitions as u64) as usize]
+                    let hash_values = batch
+                        .column_by_name("hash_value")
```
Comment (alamb) [inline datafusion/physical-plan/src/repartition/mod.rs:271]: nice

**Comment by Dandandan** [comment]
I wonder if we can make "reusing hash" an optional feature?
In some cases reuse means materializing the column (distributed), so sometimes might be better to just recompute the hash.

**Comment by jayzhan211** [comment]
> I wonder if we can make "reusing hash" an optional feature? In some cases reuse means materializing the column (distributed), so sometimes might be better to just recompute the hash.

It makes sense to me. And it is possible to make it optional

**Comment by alamb** [comment]
If it turns out that reusing hash values isn't a good idea, I think we could close the ticket as "won't do"

Another thought, similarly to what @andygrove has been exploring in https://github.com/apache/datafusion/issues/5944#issuecomment-2260794953 and I am working on in https://github.com/apache/datafusion/issues/7957 is to combine what is currently in different operators into a single one
to avoid a copy

Like maybe we could combine repartition and group by 🤔 

but that might also be a crazy idea before i have had enough ☕  this morning

**Comment by alamb** [comment]
Another point is that generating the hash value is most expensive for string columns -- so maybe we could only reuse hash values when there are string columns as part of the groupby exprs

**Comment by jayzhan211** [comment]
Let's find out if combining group + repartition is reasonable, if yes, whether to reuse hash should be trivial since we don't need to convert it to record batch anymore

**Pull request closed without merging**